torch/csrc/cuda/python_comm.cpp

```diff
@@ -13,7 +19,35 @@ void initCommMethods(PyObject *module) {
       py::call_guard<py::gil_scoped_release>())
    .def("_broadcast", [](at::Tensor& tensor, std::vector<int64_t> devices) {
      return broadcast(tensor, devices);
-   }, py::call_guard<py::gil_scoped_release>());
+   }, py::call_guard<py::gil_scoped_release>())
+   .def("_scatter", [](
+     at::Tensor& tensor,
+     std::vector<int64_t>& devices,
+     at::optional<std::vector<int64_t>> chunk_sizes,
+     int64_t dim,
+     at::optional<py::object> py_streams) {
+     at::optional<std::vector<THCStream*>> streams;
+     if (py_streams) {
+       py::handle handle = *py_streams;
+       streams = THPUtils_PySequence_to_THCStreamList(handle.ptr());
+     }
+     return scatter(tensor, devices, chunk_sizes, dim, streams);
+   },
+   py::arg("tensor"),
+   py::arg("devices"),
+   py::arg("chunk_sizes"),
+   py::arg("dim"),
+   py::arg("streams"),
+   py::call_guard<py::gil_scoped_release>())
```

Comment:
The fact that we call `THPUtils_*` from a region where we don't hold the GIL makes me nervous and is probably a bad idea. Just stick a `AutoNoGIL` right before the `return` line

Reply:
It does, it just does it too early (before the call to `THPUtils` which might use Python APIs that depend on the GIL being held!).

Reply:
Hmm now I'm getting lots of `Fatal Python error: PyEval_SaveThread: NULL tstate` errors?